dev/src/bulk-writer.ts

```diff
@@ -229,75 +241,76 @@ class BulkCommitBatch {
         results = await this.writeBatch.bulkCommit();
       } catch (err) {
         // Map the failure to each individual write's result.
-        results = [...this.pendingOps.keys()].map(path => {
-          return {key: path, writeTime: null, status: wrapError(err, stack)};
+        results = this.pendingOps.map(op => {
+          return {key: op.key, writeTime: null, status: wrapError(err, stack)};
         });
       }
       this.processResults(results);
 
-      if (this.pendingOps.size > 0) {
+      if (this.pendingOps.length > 0) {
         logger(
           'BulkWriter.bulkCommit',
           null,
           `Current batch failed at retry #${attempt}. Num failures: ` +
-            `${this.pendingOps.size}.`
+            `${this.pendingOps.length}.`
         );
 
-        this.writeBatch = new WriteBatch(this.firestore, this.writeBatch, [
-          ...this.pendingOps.keys(),
-        ]);
+        this.writeBatch = new WriteBatch(
+          this.firestore,
+          this.writeBatch,
+          this.pendingOps.map(op => op.writeBatchIndex)
+        );
       } else {
         this.completedDeferred.resolve();
         return;
       }
     }
 
-    this.failRemainingOperations(results);
+    this.processResults(results, /* failRemainingOperations= */ true);
     this.completedDeferred.resolve();
   }
 
   /**
    * Resolves the individual operations in the batch with the results.
    */
-  private processResults(results: BatchWriteResult[]): void {
-    for (const result of results) {
-      if (result.status.code === Status.OK) {
-        this.pendingOps.get(result.key)!.resolve(result);
-        this.pendingOps.delete(result.key);
+  private processResults(
+    results: BatchWriteResult[],
+    failRemainingOperations = false
```

Comment:
I think this should be "allowRetry" or something like that, which also means that you can just merge the corresponding branch with `if (!this.shouldRetry(result.status.code))`.  In the context of this method, `failRemainingOperations` is quite self-explanatory.

Reply:
Done.